osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/SliderSelectionBlueprint.cs

```diff
@@ -249,6 +251,75 @@ private void removeControlPoints(List<PathControlPoint> toRemove)
             HitObject.Position += first;
         }
 
+        private void splitControlPoints(List<PathControlPoint> toSplit)
+        {
+            // Ensure that there are any points to be split
+            if (toSplit.Count == 0)
+                return;
+
+            foreach (var c in toSplit)
+            {
+                if (c == controlPoints[0] || c == controlPoints[^1] || c.Type is null)
+                    continue;
+
+                // Split off the section of slider before this control point so the remaining control points to split are in the latter part of the slider.
+                var splitControlPoints = controlPoints.TakeWhile(current => current != c).ToList();
+
+                if (splitControlPoints.Count == 0)
+                    continue;
+
+                foreach (var current in splitControlPoints)
+                {
+                    controlPoints.Remove(current);
+                }
+
+                splitControlPoints.Add(c);
+
+                // Turn the control points which were split off into a new slider.
+                var samplePoint = (SampleControlPoint)HitObject.SampleControlPoint.DeepClone();
+                samplePoint.Time = HitObject.StartTime;
+                var difficultyPoint = (DifficultyControlPoint)HitObject.DifficultyControlPoint.DeepClone();
+                difficultyPoint.Time = HitObject.StartTime;
+
+                var newSlider = new Slider
+                {
+                    StartTime = HitObject.StartTime,
+                    Position = HitObject.Position + splitControlPoints[0].Position,
+                    NewCombo = HitObject.NewCombo,
+                    SampleControlPoint = samplePoint,
+                    DifficultyControlPoint = difficultyPoint,
+                    Samples = HitObject.Samples.Select(s => s.With()).ToList(),
+                    RepeatCount = HitObject.RepeatCount,
+                    NodeSamples = HitObject.NodeSamples.Select(n => (IList<HitSampleInfo>)n.Select(s => s.With()).ToList()).ToList(),
+                    Path = new SliderPath(splitControlPoints.Select(o => new PathControlPoint(o.Position - splitControlPoints[0].Position, o == splitControlPoints[^1] ? null : o.Type)).ToArray())
+                };
+
+                // Increase the start time of the slider before adding the new slider so the new slider is immediately inserted at the correct index and internal state remains valid.
+                HitObject.StartTime += 1;
```

Comment:
Of note, I haven't finished reviewing this because of this code. I'm very unsure of how correct this is (for instance, the `ControlPoint` creation here has slightly different `Time` specification).

If possible I'd like to see this moved into a method inside `Slider` itself or something else. It doesn't feel like this is a great place to being this, and may lead to diverging implementations of the same "clone" logic going forward.

Reply:
> the ControlPoint creation here has slightly different Time specification

I'm not sure what you mean by this. I saw that the sample control point should be 1 ms after the end of the slider, but internal updating stuff already makes it so that the sample control point is updated to this correct time. I did some testing however and noticed that the sample set of the splitted sliders do not get retained after a save and reload, so maybe something is indeed wrong with this code, but I don't know why this happens, I don't know enough about how hitsounds work in lazer. I tried to copy as much as possible from the convert-to-stream code.

> If possible I'd like to see this moved into a method inside Slider itself or something else. It doesn't feel like this is a great place to being this, and may lead to diverging implementations of the same "clone" logic going forward.

Do you think a `Slider` needs a clone method similar to `With()` in `HitSampleInfo`?

Reply:
> I did some testing however and noticed that the sample set of the splitted sliders do not get retained after a save and reload, so maybe something is indeed wrong with this code, but I don't know why this happens

I now understand why this happens and it's just `SampleControlPoint` being broken, so I'm not gonna fix that in this PR.

Reply:
> I saw that the sample control point should be 1 ms after the end of the slider, but internal updating stuff already makes it so that the sample control point is updated to this correct time

Then why are the control point times even set in [the code above](https://github.com/ppy/osu/blob/885ea4270b5bbad78a48690abdaa5f95e136b5c2/osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/SliderSelectionBlueprint.cs#L280-L282)? It just seems misleading as written if those lines aren't even doing anything and the values assigned there aren't even final.

Reply:
ill remove those lines